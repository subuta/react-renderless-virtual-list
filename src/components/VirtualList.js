import React from 'react'

import _ from 'lodash'

import {
  compose,
  withStateHandlers,
  withHandlers,
  withPropsOnChange,
  withProps,
  defaultProps,
  lifecycle,
  shouldUpdate,
  pure
} from 'recompose'

import withScroll from 'src/hocs/withScroll'

import VirtualListItem from './VirtualListItem'

export const SCROLL_DIRECTION_UP = 'SCROLL_DIRECTION_UP'
export const SCROLL_DIRECTION_DOWN = 'SCROLL_DIRECTION_DOWN'

// Default value of props.
const defaults = {
  renderListContainer: (props) => {
    const {
      className = 'c-virtual-list-container',
      children,
      style,
      setScrollContainerRef
    } = props

    return (
      <div ref={setScrollContainerRef} className={className} style={style}>{children}</div>
    )
  },

  renderList: (props) => {
    const {
      className = 'c-virtual-list',
      children,
      style
    } = props

    return (
      <div className={className} style={style}>{children}</div>
    )
  },

  renderListItem: (props) => {
    const {
      // Will be used as height before render row.
      rowSize,
      size,
      onMeasure,
      reversed,
      row,
      index,
      groupHeader,
      isGroupHeader,
      startOfRows,
      top,
      bottom
    } = props

    return (
      <VirtualListItem
        key={index}
        row={row}
        index={index}
        groupHeader={groupHeader}
        onMeasure={onMeasure}
        size={size}
        defaultRowSize={rowSize}
        reversed={reversed}
        startOfRows={startOfRows}
        isGroupHeader={isGroupHeader}
        top={top}
        bottom={bottom}
      >
        {isGroupHeader ? props.renderGroupHeader : props.children}
      </VirtualListItem>
    )
  },

  overScanCount: 4,

  height: 300,

  rowSize: {
    height: 100,
    width: 100
  }
}

const enhance = compose(
  pure,
  // Specify defaults
  defaultProps(defaults),
  withPropsOnChange(
    ['rows'],
    ({ rows, groupBy }) => {
      let groupIndices = []
      let nextRows = []
      let lastGroupHeader = undefined

      _.each(rows, (row, index) => {
        let nextGroupHeader = null
        if (groupBy) {
          nextGroupHeader = groupBy({ row, index })
        }

        if (lastGroupHeader !== undefined && nextGroupHeader !== lastGroupHeader) {
          nextRows.push({ groupHeader: nextGroupHeader })
          // Keep current indices.
          groupIndices.push(nextRows.length - 1)
        }

        nextRows.push(row)
        lastGroupHeader = nextGroupHeader
      })

      return {
        groupIndices,
        rows: nextRows
      }
    }
  ),
  withPropsOnChange(
    ['onLoadMore', 'onScroll'],
    ({ onLoadMore, onScroll }) => ({
      onScroll: _.debounce(onScroll || _.noop, 300, { leading: true, trailing: false }),
      onLoadMore: _.debounce(onLoadMore || _.noop, 300, { leading: true, trailing: false })
    })
  ),
  withPropsOnChange(
    ['renderListItem'],
    ({ renderListItem }) => ({
      renderListItem: _.memoize(renderListItem, ({ index, startOfRows, size }) => `row-${index}-${startOfRows}${size.height ? `-${size.height}` : ''}`)
    })
  ),
  withProps(({ height }) => {
    if (height === 0) return { height: defaults.height }

    return {
      height: _.isNumber(height) ? height : parseFloat(height, 10),
    }
  }),
  withStateHandlers(
    ({ reversed, rows }) => {
      const fill = _.fill(new Array(rows.length), defaults.rowSize.height)
      return {
        heightCache: fill,
        totalHeight: _.sum(fill)
      }
    },
    {
      mergeHeightCache: (state) => (heightCache) => {
        const nextHeightCache = _.merge([...state.heightCache], heightCache)
        if (_.isEqual(state.heightCache, nextHeightCache)) return
        return {
          heightCache: nextHeightCache,
          totalHeight: _.sum(nextHeightCache)
        }
      }
    }
  ),
  withScroll,
  withHandlers((props) => {
    let debouncedHeightCache = []

    const mergeHeightCache = _.debounce((heightCache) => {
      props.mergeHeightCache(heightCache)
      // Clear debouncedHeightCache
      debouncedHeightCache = []
    }, 0)

    return {
      setHeightCache: () => (index, height) => {
        debouncedHeightCache[index] = height
        mergeHeightCache(debouncedHeightCache)
      }
    }
  }),
  withHandlers(() => {
    let lastScrollTop = -1
    let isAdjusted = false
    let sizeCache = []

    return {
      onMeasure: ({ setHeightCache }) => (index, size) => {
        sizeCache[index] = size
        setHeightCache(index, size.height)
      },

      handleScroll: (props) => () => {
        const {
          scrollTop = 0,
          onScroll,
          onLoadMore,
          heightCache,
          rows,
          height,
          overScanCount,
          totalHeight,
          reversed
        } = props

        let sum = 0
        let visibleIndex = { from: -1, to: 0 }

        // Start position of rows.
        let startOfRows = reversed ? (totalHeight - (scrollTop + height)) : 0

        // End position of rows.
        let endOfRows = reversed ? (startOfRows + height) : scrollTop + height

        _.takeWhile(_.times(rows.length), (i) => {
          const h = heightCache[i]
          visibleIndex.to = i

          if (visibleIndex.from === -1 && sum + h >= (reversed ? startOfRows : scrollTop)) {
            visibleIndex.from = i
            // Use current sum as start position of rows.
            startOfRows = sum
          }

          if (sum >= endOfRows) return false

          sum += h

          return true
        })

        const overScanIndex = {
          from: visibleIndex.from - overScanCount >= 0 ? visibleIndex.from - overScanCount : 0,
          to: visibleIndex.to + overScanCount <= rows.length - 1 ? visibleIndex.to + overScanCount : rows.length - 1
        }

        // Minus top overScan height for correct startOfRows.
        startOfRows -= _.sum(_.slice(heightCache, overScanIndex.from, visibleIndex.from))
        endOfRows += _.sum(_.slice(heightCache, visibleIndex.to, overScanIndex.to))

        const direction = lastScrollTop > scrollTop ? SCROLL_DIRECTION_UP : SCROLL_DIRECTION_DOWN

        lastScrollTop = scrollTop

        const hasInitialized = !reversed || isAdjusted
        const isEdge = overScanIndex.to >= rows.length - 1

        // Call onScroll.
        if (onScroll && hasInitialized) {
          onScroll({
            scrollTop,
            direction,
            visibleIndex,
            overScanIndex
          })
        }

        if (onLoadMore && hasInitialized && isEdge) {
          onLoadMore({
            scrollTop,
            direction,
            visibleIndex,
            overScanIndex
          })
        }

        return {
          visibleIndex,
          overScanIndex,
          positions: {
            from: startOfRows,
            to: endOfRows
          }
        }
      },

      getSizeCache: () => () => sizeCache,

      getIsAdjusted: () => () => isAdjusted,

      getStyles: ({ totalHeight, heightCache, height }) => () => {
        const containerStyle = {
          height,
          width: '100vw',
          overflowX: 'auto',
          willChange: 'transform'
        }

        const listStyle = {
          position: 'relative',
          minHeight: '100%',
          width: '100%',
          height: totalHeight
        }

        return {
          containerStyle,
          listStyle
        }
      },

      adjustScrollTop: (props) => (scrollTop) => {
        const {
          requestScrollTo,
          requestScrollToBottom,
          hasScrolledToBottom
        } = props

        if (isAdjusted) {
          requestScrollTo(scrollTop)
        } else {
          requestScrollToBottom()
          requestAnimationFrame(() => {
            isAdjusted = hasScrolledToBottom()
          })
        }
      }
    }
  }),
  withPropsOnChange(
    ['scrollTop', 'height', 'totalHeight'],
    (props) => ({
      ...props.handleScroll(),
      ...props.getStyles()
    })
  ),
  withPropsOnChange(
    ['rows'],
    (props) => {
      const {
        rows,
        heightCache,
        mergeHeightCache
      } = props

      const fill = _.fill(new Array(rows.length), defaults.rowSize.height)
      mergeHeightCache(_.merge(fill, heightCache))
    }
  ),
  lifecycle({
    componentDidMount () {
      if (this.props.reversed) {
        this.props.adjustScrollTop()
      }
    },

    getSnapshotBeforeUpdate (prevProps) {
      if (!_.isEqual(prevProps.rows, this.props.rows)) {
        return prevProps.scrollTop
      }

      if (prevProps.totalHeight !== this.props.totalHeight) {
        const diff = this.props.totalHeight - prevProps.totalHeight
        return this.props.scrollTop + diff
      }

      return null
    },

    componentDidUpdate (prevProps, prevState, scrollTop) {
      if (!this.props.reversed) return

      if (scrollTop !== null) {
        this.props.adjustScrollTop(scrollTop)
      }
    }
  }),
  shouldUpdate((props, nextProps) => {
    const isOverScanIndexChanged = !_.isEqual(props.overScanIndex, nextProps.overScanIndex)
    const isHeightChanged = !_.isEqual(props.height, nextProps.height)
    const isTotalHeightChanged = !_.isEqual(props.totalHeight, nextProps.totalHeight)

    return isOverScanIndexChanged || isHeightChanged || isTotalHeightChanged
  })
)

export default enhance((props) => {
  const {
    rows = [],
    containerStyle = {},
    listStyle = {},
    groupIndices,
    // Will be used as height before render row.
    overScanIndex = {
      from: 0,
      to: rows.length - 1
    },
    heightCache = [],
    renderList,
    setScrollContainerRef,
    renderListItem,
    renderListContainer,
    getSizeCache,
    totalHeight,
    reversed
  } = props

  let startOfRows = _.get(props, 'positions.from')
  let oppositeEdge = 0
  const endOfRows = _.get(props, 'positions.to')

  const List = renderList
  const Container = renderListContainer
  const sizeCache = getSizeCache()

  return (
    <Container
      style={containerStyle}
      setScrollContainerRef={setScrollContainerRef}
    >
      <List style={listStyle}>
        {_.map(rows, (row, index) => {
          const isGroupHeader = _.includes(groupIndices, index)

          if (!isGroupHeader) {
            // No-render if index out of range.
            if (index < overScanIndex.from || index > overScanIndex.to) return null
            if (startOfRows > endOfRows) return null
          }

          const size = sizeCache[index] || {}

          let itemProps = {
            ...props,
            row,
            index,
            size,
            startOfRows,
            [reversed ? 'bottom' : 'top']: startOfRows,
            isGroupHeader
          }

          startOfRows += heightCache[index]

          if (isGroupHeader) {
            const oppositeEdgeAttr = reversed ? 'top' : 'bottom'

            itemProps = {
              ...itemProps,
              [oppositeEdgeAttr]: totalHeight - itemProps[reversed ? 'bottom' : 'top'] + 8,
              [reversed ? 'bottom' : 'top']: oppositeEdge
            }

            console.log('itemProps = ', itemProps)

            oppositeEdge = startOfRows
          }

          return renderListItem(itemProps)
        })}
      </List>
    </Container>
  )
})
