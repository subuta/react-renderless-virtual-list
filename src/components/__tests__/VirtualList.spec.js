import React from 'react'
import { mount } from 'enzyme'
import sinon from 'sinon'
import { create } from 'react-test-renderer'

import _ from 'lodash'

import VirtualList from 'src/components/VirtualList'

let clock

import ResizeObserver from 'resize-observer-polyfill'

const mockedObserve = jest.fn()
const mockedUnobserve = jest.fn()

const mockedRequestScrollTo = jest.fn()

// Mock withScroll while testing.
jest.mock('src/hocs/withScroll', () => {
  return jest.fn().mockImplementation((Component) => {
    return (props) => (
      <Component
        {...props}
        requestScrollTo={mockedRequestScrollTo}
      />
    )
  })
})

// Mock for simulate ResizeObserver at jest(with jsdom) environment.
jest.mock('resize-observer-polyfill', () => {
  return jest.fn().mockImplementation(() => {
    return {
      observe: mockedObserve,
      unobserve: mockedUnobserve
    }
  })
})

beforeEach(() => {
  clock = sinon.useFakeTimers()
  mockedRequestScrollTo.mockClear()
  ResizeObserver.mockClear()
  mockedObserve.mockClear()
  mockedUnobserve.mockClear()
})

afterEach(() => {
  clock.restore()
})

test('should not render with empty rows.', () => {
  const child = sinon.spy(() => <span>hoge</span>)

  mount(
    <VirtualList
      rows={[]}
    >
      {child}
    </VirtualList>
  )

  // Should not render anything for empty rows.
  expect(child.callCount).toBe(0)
})

test('should render with 1 rows.', () => {
  const child = sinon.spy(({ row, index, style }) => {
    return (
      <span className={`row-${index + 1}`} style={style}>{row.id}</span>
    )
  })

  const rows = _.times(1, (n) => ({ id: n + 1 }))

  const tree = create(
    <VirtualList
      rows={rows}
    >
      {child}
    </VirtualList>
  )

  expect(child.callCount).toBe(1)

  const props = child.firstCall.args[0]

  expect(props.row).toEqual({ id: 1 })
  expect(props.index).toEqual(0)
  expect(props.setSizeRef).toBeInstanceOf(Function)

  // Testing for snapshot.
  expect(tree.toJSON()).toMatchSnapshot()
})

test('should render with height as px', () => {
  const child = sinon.spy(({ row, index, style }) => {
    return (
      <span className={`row-${index + 1}`} style={style}>{row.id}</span>
    )
  })

  const rows = _.times(1, (n) => ({ id: n + 1 }))

  const tree = create(
    <VirtualList
      height="400px"
      rows={rows}
    >
      {child}
    </VirtualList>
  )

  expect(child.callCount).toBe(1)

  const props = child.firstCall.args[0]

  expect(props.row).toEqual({ id: 1 })
  expect(props.index).toEqual(0)
  expect(props.setSizeRef).toBeInstanceOf(Function)

  // Testing for snapshot.
  expect(tree.toJSON()).toMatchSnapshot()
})

test('should re-render at onMeasure call at VirtualListItem', () => {
  const child = sinon.spy(({ row, index, style }) => {
    return (
      <span className={`row-${index + 1}`} style={style}>{row.id}</span>
    )
  })

  const rows = _.times(1, (n) => ({ id: n + 1 }))

  let simulateResize = _.noop

  ResizeObserver.mockImplementationOnce(cb => {
    simulateResize = cb
    // Delay execution for retrieve ref.
    requestAnimationFrame(() => cb([
      {
        contentRect: {
          top: 0,
          right: 200,
          bottom: 200,
          left: 0,
          height: 100,
          width: 100
        }
      }
    ]))
    return {
      observe: mockedObserve,
      unobserve: mockedUnobserve
    }
  })

  mount(
    <VirtualList
      rows={rows}
    >
      {child}
    </VirtualList>
  )

  expect(child.callCount).toBe(1)

  let props = child.firstCall.args[0]

  expect(props.style).toEqual({
    position: 'absolute',
    top: 0,
    left: 0,
    minHeight: 100,
    minWidth: 100
  })

  // At first resize.
  clock.runAll()

  expect(child.callCount).toBe(2)

  props = child.secondCall.args[0]

  expect(props.style).toEqual({
    position: 'absolute',
    top: 0,
    left: 0,
    minHeight: 100,
    minWidth: 100,
    height: 200,
    width: 200
  })

  // simulate resize event of element.
  simulateResize([
    {
      contentRect: {
        top: 0,
        right: 300,
        bottom: 300,
        left: 0,
        height: 100,
        width: 100
      }
    }
  ])

  clock.runAll()

  expect(child.callCount).toBe(3)

  props = child.thirdCall.args[0]

  expect(props.style).toEqual({
    position: 'absolute',
    top: 0,
    left: 0,
    minHeight: 100,
    minWidth: 100,
    height: 300,
    width: 300
  })
})

test('should render child as pure.', () => {
  const renderListContainer = sinon.spy(({ children }) => <div>{children}</div>)
  const renderList = sinon.spy(({ children }) => <div>{children}</div>)

  const child = sinon.spy(({ row, index, style }) => {
    return (
      <span className={`row-${index + 1}`} style={style}>{row.id}</span>
    )
  })

  const rows = _.times(1, (n) => ({ id: n + 1 }))
  const initialProps = {
    rows,
    renderListContainer,
    renderList
  }

  const wrapper = mount(
    <VirtualList
      {...initialProps}
    >
      {child}
    </VirtualList>
  )

  expect(renderListContainer.callCount).toBe(1)
  expect(renderList.callCount).toBe(1)
  expect(child.callCount).toBe(1)

  const props = child.firstCall.args[0]

  expect(props.row).toEqual({ id: 1 })
  expect(props.index).toEqual(0)
  expect(props.setSizeRef).toBeInstanceOf(Function)

  // Should not re-render if props not changed.
  wrapper.setProps(initialProps)

  expect(renderListContainer.callCount).toBe(1)
  expect(renderList.callCount).toBe(1)
  expect(child.callCount).toBe(1)

  // TODO: should memoize render child.
  // Should not re-render if scrollTop changed.
  wrapper.setProps({
    ...initialProps,
    scrollTop: 100
  })

  expect(renderListContainer.callCount).toBe(2)
  expect(renderList.callCount).toBe(2)
  expect(child.callCount).toBe(1)
})

test('should render with 30 rows.', () => {
  const renderListContainer = sinon.spy(({ children, style }) => <div style={style}>{children}</div>)
  const renderList = sinon.spy(({ children, style }) => <div style={style}>{children}</div>)

  const child = sinon.spy(({ row, index, style }) => {
    return (
      <span className={`row-${index + 1}`} style={style}>{row.id}</span>
    )
  })

  const count = 30
  const rows = _.times(count, (n) => ({ id: n + 1 }))
  const defaultRowHeight = 100
  const totalHeight = defaultRowHeight * count

  const tree = create(
    <VirtualList
      renderListContainer={renderListContainer}
      renderList={renderList}
      rows={rows}
      defaultRowHeight={defaultRowHeight}
    >
      {child}
    </VirtualList>
  )

  expect(renderListContainer.callCount).toBe(1)

  const listContainerProps = renderListContainer.firstCall.args[0]
  expect(listContainerProps.style).toEqual({
    height: 300,
    width: '100vw',
    overflowX: 'auto',
    willChange: 'transform'
  })

  // Should render list.
  expect(renderList.callCount).toBe(1)

  const listProps = renderList.firstCall.args[0]
  expect(listProps.style).toEqual({
    position: 'relative',
    minHeight: '100%',
    width: '100%',
    height: totalHeight
  })

  // Should render only-visible child rows.
  expect(child.callCount).toBe(7)

  clock.runAll()

  // Should not-render after setDebouncedHeightCache.
  expect(child.callCount).toBe(7)

  const childProps = child.firstCall.args[0]

  expect(childProps.row).toEqual({ id: 1 })
  expect(childProps.index).toEqual(0)
  expect(childProps.setSizeRef).toBeInstanceOf(Function)
  expect(childProps.style).toEqual({
    position: 'absolute',
    top: 0,
    left: 0,
    minHeight: 100,
    minWidth: 100
  })

  const lastCallChildProps = child.lastCall.args[0]

  expect(lastCallChildProps.row).toEqual({ id: 7 })
  expect(lastCallChildProps.index).toEqual(6)
  expect(lastCallChildProps.setSizeRef).toBeInstanceOf(Function)
  expect(lastCallChildProps.style).toEqual({
    position: 'absolute',
    top: 600,
    left: 0,
    minHeight: 100,
    minWidth: 100
  })

  // requestScrollTo should called at componentDidMount
  expect(mockedRequestScrollTo).not.toHaveBeenCalled()

  // Testing for snapshot.
  expect(tree.toJSON()).toMatchSnapshot()
})

test('should render with reversed 30 rows.', () => {
  const renderListContainer = sinon.spy(({ children }) => <div>{children}</div>)
  const renderList = sinon.spy(({ children }) => <div>{children}</div>)

  const child = sinon.spy(({ row, index, style }) => {
    return (
      <span className={`row-${index + 1}`} style={style}>{row.id}</span>
    )
  })

  const count = 30
  const rows = _.times(count, (n) => ({ id: n + 1 }))
  const defaultRowHeight = 100
  const totalHeight = defaultRowHeight * count

  const tree = create(
    <VirtualList
      renderListContainer={renderListContainer}
      renderList={renderList}
      rows={rows}
      defaultRowHeight={defaultRowHeight}
      reversed
    >
      {child}
    </VirtualList>
  )

  expect(renderListContainer.callCount).toBe(1)

  const listContainerProps = renderListContainer.firstCall.args[0]
  expect(listContainerProps.style).toEqual({
    height: 300,
    width: '100vw',
    overflowX: 'auto',
    willChange: 'transform'
  })

  // Should render list.
  expect(renderList.callCount).toBe(1)

  const listProps = renderList.firstCall.args[0]
  expect(listProps.style).toEqual({
    position: 'relative',
    minHeight: '100%',
    width: '100%',
    height: totalHeight
  })

  // Should render only-visible child rows.
  expect(child.callCount).toBe(7)

  clock.runAll()

  // Should not-render after setDebouncedHeightCache.
  expect(child.callCount).toBe(7)

  const childProps = child.firstCall.args[0]

  expect(childProps.row).toEqual({ id: 24 })
  expect(childProps.index).toEqual(23)
  expect(childProps.setSizeRef).toBeInstanceOf(Function)
  expect(childProps.style).toEqual({
    position: 'absolute',
    bottom: 2300,
    left: 0,
    minHeight: 100,
    minWidth: 100
  })

  const lastCallChildProps = child.lastCall.args[0]

  expect(lastCallChildProps.row).toEqual({ id: 30 })
  expect(lastCallChildProps.index).toEqual(29)
  expect(lastCallChildProps.setSizeRef).toBeInstanceOf(Function)
  expect(lastCallChildProps.style).toEqual({
    position: 'absolute',
    bottom: 2900,
    left: 0,
    minHeight: 100,
    minWidth: 100
  })

  // requestScrollTo should called at componentDidMount
  expect(mockedRequestScrollTo).toHaveBeenCalledWith(totalHeight)

  // Testing for snapshot.
  expect(tree.toJSON()).toMatchSnapshot()
})
