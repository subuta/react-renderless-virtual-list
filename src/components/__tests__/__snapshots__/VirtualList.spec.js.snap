// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with 1 rows. 1`] = `
<div
  className="c-virtual-list-container"
  style={
    Object {
      "height": 300,
      "overflowX": "auto",
      "width": "100vw",
      "willChange": "transform",
    }
  }
>
  <div
    className="c-virtual-list"
    style={
      Object {
        "height": 10000000,
        "minHeight": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <span
      className="row-1"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 0,
        }
      }
    >
      1
    </span>
  </div>
</div>
`;

exports[`should render with 30 rows. 1`] = `
<div
  style={
    Object {
      "height": 300,
      "overflowX": "auto",
      "width": "100vw",
      "willChange": "transform",
    }
  }
>
  <div
    style={
      Object {
        "height": 10000000,
        "minHeight": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <span
      className="row-1"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 0,
        }
      }
    >
      1
    </span>
    <span
      className="row-2"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 100,
        }
      }
    >
      2
    </span>
    <span
      className="row-3"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 200,
        }
      }
    >
      3
    </span>
    <span
      className="row-4"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 300,
        }
      }
    >
      4
    </span>
    <span
      className="row-5"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 400,
        }
      }
    >
      5
    </span>
    <span
      className="row-6"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 500,
        }
      }
    >
      6
    </span>
  </div>
</div>
`;

exports[`should render with height as px 1`] = `
<div
  className="c-virtual-list-container"
  style={
    Object {
      "height": 400,
      "overflowX": "auto",
      "width": "100vw",
      "willChange": "transform",
    }
  }
>
  <div
    className="c-virtual-list"
    style={
      Object {
        "height": 10000000,
        "minHeight": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <span
      className="row-1"
      style={
        Object {
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
          "top": 0,
        }
      }
    >
      1
    </span>
  </div>
</div>
`;

exports[`should render with reversed 30 rows. 1`] = `
<div>
  <div>
    <span
      className="row-1"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
        }
      }
    >
      1
    </span>
    <span
      className="row-2"
      style={
        Object {
          "bottom": 100,
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
        }
      }
    >
      2
    </span>
    <span
      className="row-3"
      style={
        Object {
          "bottom": 200,
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
        }
      }
    >
      3
    </span>
    <span
      className="row-4"
      style={
        Object {
          "bottom": 300,
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
        }
      }
    >
      4
    </span>
    <span
      className="row-5"
      style={
        Object {
          "bottom": 400,
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
        }
      }
    >
      5
    </span>
    <span
      className="row-6"
      style={
        Object {
          "bottom": 500,
          "left": 0,
          "minHeight": 100,
          "position": "absolute",
        }
      }
    >
      6
    </span>
  </div>
</div>
`;
